static/js/Promotions.jsx

```diff
@@ -1,200 +1,221 @@
-        return `<div id="bannerTextBox">
-	<span class="${context.interfaceLang === "hebrew" ? "int-he" : "int-en" }" style="font-weight: bold">
-        ${matchingAd.bodyText}
-    </span>
-</div>
-<div id="bannerButtonBox">
-	<a class="button white ${context.interfaceLang === "hebrew" ? "int-he" : "int-en" }" 
-	    href="${matchingAd.buttonUrl}"   
-	    onclick='gtag("event", "promo_clicked", {  
-            campaignID: matchingAd.campaignId, adType:matchingAd.adType
-            })' 
-        target="_blank">
-        <span>${matchingAd.buttonText}</span>
-    </a>
-</div>`
-    }
-
-    function styleAds() {
-        if (adType === "banner") {
-            const matchingAd = matchingAds[0] // Only allow a single banner
-            if (!matchingAd) {return null}
-            const bannerHtml = createBannerHtml(matchingAd);
-            return <InterruptingMessage
-            messageName={matchingAd.campaignId}
-            messageHTML={bannerHtml}
-            style="banner"
-            repetition={matchingAd.repetition}
-            onClose={rerender} />
-        } else {
-              const sidebarAds = matchingAds.map(ad => <SidebarAd matchingAd={ad} key={ad.campaignId} />);
-              return (sidebarAds)
-        }
-    }
+  return matchingAds
+    ? matchingAds.map((ad) => (
+        <SidebarAd context={context} matchingAd={ad} key={ad.campaignId} />
+      ))
+    : null;
+};
 
-    return matchingAds ? styleAds() : null
+function trackSidebarAdImpression(ad) {
+  console.log(ad.campaignId + " has been seen");
+  gtag("event", "promo_viewed", {
+    campaignID: ad.campaignId,
+    adType: "sidebar",
+  });
+}
 
+function trackSidebarAdClick(ad) {
+  gtag("event", "promo_clicked", {
+    campaignID: ad.campaignId,
+    adType: "sidebar",
+  });
 }
 
-const SidebarAd = ({matchingAd}) => {
-    const classes = classNames({
-        sidebarPromo: 1,
-        blue: matchingAd.hasBlueBackground,
-    })
+// Don't continuously rerender a SidebarAd if the parent component decides to rerender
+// This is done to prevent multiple views from registering from OnInView
+const SidebarAd = React.memo(({ context, matchingAd }) => {
+  const classes = classNames({
+    sidebarPromo: 1,
+    blue: matchingAd.hasBlueBackground,
+  });
 
-    function getButton() {
-        return <a className={matchingAd.buttonStyle} href={matchingAd.buttonUrl}
-                  onClick={() => gtag("event", "promo_clicked", {
-                              campaignID: matchingAd.campaignId,
-                              adType:matchingAd.adType
-                            })}>
-        <img src={`/static/icons/${matchingAd.buttonIcon}`} aria-hidden="true" />
-        {matchingAd.buttonText}</a>
-    }
+  function getButton() {
+    return (
+      <a
+        className="button small"
+        href={matchingAd.buttonURL}
+        onClick={() => trackSidebarAdClick(matchingAd)}
+      >
+        {matchingAd.buttonIcon?.data ? (
+          <img
```

Comment:
The media files are being served as assets on the Strapi server. The person who creates the sidebar ad might not add an icon for the button

Reply:
I think my question should be clarified because unfortunately, not sure your answer helped clarify for me.  
I'm asking regarding your comments in the code of writing an additional "middleware" piece to serve media assets? 
Why? In general writing code that needs to be environment aware is not the best practice. In other parts of the applications and on the server we write code that serves media from a predefined (potentially by a variable that can be configured) place. The code is then agnostic of it and just serves from where its told. This saves potential complications and wires crossed later on. 
Why is there a need for manual handling of both absolute and relative paths here? I think that part needs thinking about.